File: macosx/ExpandedPathToIconTransformer.m

#import "ExpandedPathToIconTransformer.h"

@implementation ExpandedPathToIconTransformer

+ (Class) transformedValueClass
{
    return [NSImage class];
}

+ (BOOL) allowsReverseTransformation
{
    return NO;
}

- (id) transformedValue: (id) value
{
    if (!value)
        return nil;
    
    NSString * path = [value stringByExpandingTildeInPath];
    NSImage * icon;
    //show a folder icon if the folder doesn't exist
    if (![[NSFileManager defaultManager] fileExistsAtPath: path] && [[path pathExtension] isEqualToString: @""])
        icon = [[NSWorkspace sharedWorkspace] iconForFileType: NSFileTypeForHFSTypeCode('fldr')];
    else
        icon = [[NSWorkspace sharedWorkspace] iconForFile: [value stringByExpandingTildeInPath]];
    
    [icon setScalesWhenResized: YES];
    [icon setSize: NSMakeSize(16.0, 16.0)];
    
    return icon;
}

@end

File: macosx/ExpandedPathToIconTransformer.swift
// This file Copyright © 2007-2023 Transmission authors and contributors.
// It may be used under the MIT (SPDX: MIT) license.
// License text can be found in the licenses/ folder.

import AppKit

class ExpandedPathToIconTransformer: ValueTransformer {
    override class func transformedValueClass() -> AnyClass {
        return NSImage.self
    }

    override class func allowsReverseTransformation() -> Bool {
        return false
    }

    override func transformedValue(_ value: Any?) -> Any? {
        guard let value = value as? String else {
            return nil
        }

        let path = (value as NSString).expandingTildeInPath
        let icon: NSImage
        // show a folder icon if the folder doesn't exist
        if (path as NSString).pathExtension.isEmpty,
           !FileManager.default.fileExists(atPath: path) {
            icon = NSWorkspace.shared.icon(forFileType: NSFileTypeForHFSTypeCode(OSType(kGenericFolderIcon)))
        } else {
            icon = NSWorkspace.shared.icon(forFile: path)
        }
        icon.size = NSSize(width: 16.0, height: 16.0)

        return icon
    }
}
